src/pipeline/rackhd/source_code/UnitTest.groovy

```diff
@@ -0,0 +1,78 @@
+package pipeline.rackhd.source_code
+
+def testWithSudo(String repo_name, String repo_dir){
+    try{
+        withCredentials([
+            usernamePassword(credentialsId: 'ff7ab8d2-e678-41ef-a46b-dd0e780030e1',
+                passwordVariable: 'SUDO_PASSWORD',
+                usernameVariable: 'SUDO_USER')
+        ]){
+            sh """#!/bin/bash -ex
+            trap 'echo $SUDO_PASSWORD |sudo -S chown -R $USER:$USER $repo_dir' SIGINT SIGTERM SIGKILL EXIT
+            pushd $repo_dir
+            echo $SUDO_PASSWORD |sudo -S ./HWIMO-TEST
+            popd
+            """
+        }
+    } finally{
+        dir("$WORKSPACE"){
+            archiveArtifacts "$repo_name/*.xml"
+            junit "$repo_name/*.xml"
+        }
+    }
+}
+
+def testWithoutSudo(String repo_name, String repo_dir){
+    try{
+        sh """#!/bin/bash -ex
+        pushd $repo_dir
+        ./HWIMO-TEST
+        popd
+        """
+    } finally{
+        dir("$WORKSPACE"){
+            archiveArtifacts "$repo_name/*.xml"
+            junit "$repo_name/*.xml"
+        }
+    }
+}
+
+def runTest(String repo_name, Map manifest_dict, ArrayList<String> used_resources, boolean with_sudo){
+    /*
+    :manifest_dict is a map contains manifest stash name and stash path:
+    :repo_name is a String which is the name of the repository 
+    :used_resources is a array which records the used resources
+    */
+    def share_method = new pipeline.common.ShareMethod()
+    def manifest = new pipeline.common.Manifest()
+    String label_name="unittest"
+    try{
+        lock(label:label_name,quantity:1){
+            node_name = share_method.occupyAvailableLockedResource(label_name, used_resources)
+            node(node_name){
+                deleteDir()
+                String library_dir = "$WORKSPACE/on-build-config"
+                share_method.checkoutOnBuildConfig(library_dir)
+                String manifest_path = manifest.unstashManifest(manifest_dict, "$WORKSPACE")
+                String repo_dir = "$WORKSPACE/$repo_name"
```

Comment:
you should tell the user of this function:  the repo should be cloned in advanced to $WORKSPACE/$repo_name, so that runTest() can step into that folder and run HWIMO-TEST.
otherwise , it makes user very difficult to understand the pre-condition to use this function.
actually, the repo_dir can become an input  parameter if you like